oeps/oep-0038-Data-Modeling.rst

```diff
@@ -0,0 +1,411 @@
+      potential downtime.
+
+Due to the potentially high cost of changing a model, it pays to get it
+as correct as possible the first time.
+
+Think about the person who will analyze this data later and the person
+who manages the system day to day. (That person could possibly be future
+you!) Try to do as much careful design up front to make your quality of
+life better later.
+
+History is best (but events are a close second)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Historical data accuracy is best when history about a change is captured at the database level. Changes in data relating to financial systems (e.g. enrollments, payments, course price changes) should be stored historically where the change is made, preferably in the same system as the system of record. This allows us to reconstruct the data at different points with much greater reliability.
+
+If, for some reason, it is not possible to create history at the time of
+entry (for example, if we expect the table to become too large or the
+writes to be too frequent), it is recommended that an event be issued in its
+place.
+
+
+Data Modeling Standards
+-----------------------
+
+The standards below are designed to ensure edX can gain the highest value and insights from the data.
+The application of these standards is the most basic level of support to which all applications in the Open edX ecosystem
+should adhere. When creating new applications or models please ensure the models being created conform to the following.
+
+Use a numeric primary key
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+-  It is recommended to use :code:`BigAutoField`_.
+.. _BigAutoField: https://docs.djangoproject.com/en/2.2/ref/models/fields/#bigautofield
+
+-  Do not use composite based primary keys. Use a primary key column.
+
+Have updated and created timestamps
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+-  The preferred method for doing this in Open edX Django applications is to inherit the :code:`TimeStampedModel`_. class.
+.. _TimeStampedModel: https://django-model-utils.readthedocs.io/en/latest/models.html#timestampedmodel
+
+- Time should be stored in UTC time by setting :code:`USE_TZ=True`_. in your python config.
+.. _USE_TZ=True: https://docs.djangoproject.com/en/2.2/topics/i18n/timezones/#overview
+
+-  If for some reason you can not inherit from :code:`TimeStampedModel`_. use the following naming conventions:
+
+   -  Created date should be named: “created”
+
+   -  Updated date should be named: “modified”
+
+Foreign keys
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+-  Data should be joined using primary keys wherever possible
+
+-  Foreign keys should use a naming convention of :code:`<object_name>_id` where object name is the name of the table of the foreign key relationship.
+
+-  Do not join on things such as username, email address, or other dimensions of data that may change over time
+
+-  Do not join on PII
+
+- Joining between IDAs should be done by using a universally unique identifier (UUID)
```

Comment:
Is this line in conflict with the first line in this section? If not, perhaps it should be clarified when joining with primary keys is the best practice vs. joining with UUIDs?

Reply:
UUIDs can span databases, foreign keys cannot.